File: internal/limiter/tokenbucket.go
package limiter

import (
	"context"
	_ "embed"
	"time"

	"github.com/redis/go-redis/v9"
)

//go:embed scripts/token_bucket.lua
var tokenBucketSrc string

//go:embed scripts/token_bucket_reconcile.lua
var tokenBucketReconcileSrc string

// The Lua sources are loaded once and run by SHA (EVALSHA, with an EVAL
// fallback) on each call, so the atomic read-refill-debit-write happens
// server-side in one round-trip — the whole point of STEP 4.
var (
	tokenBucketScript    = redis.NewScript(tokenBucketSrc)
	tokenBucketReconcile = redis.NewScript(tokenBucketReconcileSrc)
)

// TokenBucket is the default algorithm (build-order STEP 4). It natively
// supports weighted cost, which is what makes it the right fit for
// variable-cost LLM requests. Unlike the in-memory Bucket (step 2) or the
// INCRBY FixedWindow (step 3), read -> lazy-refill -> conditional debit ->
// write all run inside ONE Lua script, so N stateless replicas share a single
// correct budget with no read-modify-write race.
type TokenBucket struct {
	rdb      redis.Cmdable // *redis.Client or *redis.ClusterClient
	capacity float64       // burst ceiling
	refill   float64       // tokens per second (the sustained rate)
	ttl      time.Duration // idle buckets auto-expire
}

func NewTokenBucket(rdb redis.Cmdable, capacity, refillPerSec float64, ttl time.Duration) *TokenBucket {
	if ttl <= 0 {
		ttl = time.Hour
	}
	return &TokenBucket{rdb: rdb, capacity: capacity, refill: refillPerSec, ttl: ttl}
}

func (t *TokenBucket) Allow(ctx context.Context, key string, cost float64) (Decision, error) {
	now := time.Now()
	_ = now

	// HINT 1 — run the script atomically. redis.Script.Run does EVALSHA (falling
	// back to EVAL on NOSCRIPT), so the entire bucket update is one race-free
	// round-trip. KEYS[1] is the bucket key; the ARGV order MUST match the header
	// contract in scripts/token_bucket.lua exactly:
	//     res, err := tokenBucketScript.Run(ctx, t.rdb,
	//         []string{key},
	//         t.capacity, t.refill, cost, now.UnixMilli(), t.ttl.Milliseconds(),
	//     ).Result()

	// HINT 2 — errors bubble up. On err return (Decision{}, err) and let the
	// gateway apply the fail-open/closed policy (ADR-008); don't invent a
	// decision here.

	// HINT 3 — parse. The script returns { allowed(0|1), remaining(string),
	// reset_ms }. parseDecision (limiter.go) turns that array into a Decision and
	// converts reset_ms into ResetAt relative to `now`:
	//     return parseDecision(res, t.capacity, now)

	return Decision{}, nil // TODO: implement via tokenBucketScript.Run + parseDecision
}

// Reconcile settles the optimistic pre-flight debit: delta = actual - estimate.
func (t *TokenBucket) Reconcile(ctx context.Context, key string, delta float64) error {
	// TODO: run tokenBucketReconcile with KEYS = [key] and ARGV = capacity,
	// delta, ttl_ms (see scripts/token_bucket_reconcile.lua). delta > 0 consumes
	// more, delta < 0 refunds; the script clamps to [0, capacity]. Return .Err().
	//     return tokenBucketReconcile.Run(ctx, t.rdb,
	//         []string{key}, t.capacity, delta, t.ttl.Milliseconds()).Err()
	return nil // TODO
}

File: internal/limiter/scripts/token_bucket.lua
-- Token bucket, executed atomically by Redis (no read-modify-write race across gateways).
-- KEYS[1] = bucket key (hash with fields: tokens, ts)
-- ARGV[1] = capacity        (max tokens / burst)
-- ARGV[2] = refill          (tokens per second)
-- ARGV[3] = cost            (weight of THIS request: 1, or estimated LLM tokens, or dollars)
-- ARGV[4] = now_ms          (caller clock, ms)
-- ARGV[5] = ttl_ms          (auto-expire idle buckets)
-- returns { allowed(0|1), remaining(string), reset_ms }

local capacity = tonumber(ARGV[1])
local refill   = tonumber(ARGV[2])
local cost     = tonumber(ARGV[3])
local now      = tonumber(ARGV[4])
local ttl      = tonumber(ARGV[5])

-- TODO (STEP 4): implement the atomic read -> lazy-refill -> conditional debit
-- -> write. This is THE distributed-correctness win: all of it runs server-side
-- so concurrent gateways can't interleave a stale read with a write.
--
-- HINT 1 — read current state with HMGET KEYS[1] 'tokens' 'ts'. On a cold bucket
--          (tokens == nil) start FULL: tokens = capacity, ts = now (the
--          cold-start burst trade-off, ADR-005).
local state = redis.call("HMGET", KEYS[1], 'tokens', 'ts')
local tokens = tonumber(state[1])
local ts = tonumber(state[2])
if tokens == nil then
    tokens = capacity
    ts = now
end
-- HINT 2 — lazy refill by elapsed time, capped at capacity:
local elapsed = math.max(0, now - ts) / 1000.0
tokens = math.min(capacity, tokens + elapsed * refill)
-- HINT 3 — conditional debit: only when tokens >= cost do you set allowed = 1
--          and tokens = tokens - cost; otherwise allowed = 0 and don't spend.
local allowed = 0
local reset_ms = 0
if tokens >= cost then
    allowed = 1
    tokens = tokens - cost
else
    if refill > 0 then
        reset_ms = math.ceil((cost - tokens) / refill * 1000)
    else
        reset_ms = 0
    end
end
-- HINT 4 — persist: HSET KEYS[1] 'tokens' tokens 'ts' now, then
--          PEXPIRE KEYS[1] ttl so idle buckets disappear.
redis.call("HSET", KEYS[1], "tokens", tokens, "ts", now)
redis.call("PEXPIRE", KEYS[1], ttl)
-- HINT 5 — reset_ms for Retry-After: when short, ceil((cost - tokens)/refill*1000)
--          (guard refill > 0); otherwise 0.
-- HINT 6 — return { allowed, tostring(tokens), reset_ms }. tokens is stringified
--          so fractional precision survives (parseDecision reads it back).
return { allowed, tostring(tokens), reset_ms }
